pinot-minion/src/main/java/com/linkedin/pinot/minion/executor/SegmentConversionExecutor.java

```diff
@@ -0,0 +1,121 @@
+    String downloadURL = configs.get(MinionConstants.DOWNLOAD_URL_KEY);
+    String uploadURL = configs.get(MinionConstants.UPLOAD_URL_KEY);
+    String originalSegmentCrc = configs.get(MinionConstants.ORIGINAL_SEGMENT_CRC_KEY);
+
+    LOGGER.info("Start executing {} on table: {}, segment: {} with downloadURL: {}, uploadURL: {}", taskType, tableName,
+        segmentName, downloadURL, uploadURL);
+
+    File tempDataDir = new File(new File(MINION_CONTEXT.getDataDir(), taskType), "tmp-" + System.nanoTime());
+    Preconditions.checkState(tempDataDir.mkdirs());
+    try {
+      // Download the tarred segment file
+      File tarredSegmentFile = new File(tempDataDir, "tarredSegmentFile");
+      SegmentFetcherFactory.getSegmentFetcherBasedOnURI(downloadURL)
+          .fetchSegmentToLocal(downloadURL, tarredSegmentFile);
+
+      // Un-tar the segment file
+      File segmentDir = new File(tempDataDir, "segmentDir");
+      TarGzCompressionUtils.unTar(tarredSegmentFile, segmentDir);
+      File[] files = segmentDir.listFiles();
+      Preconditions.checkState(files != null && files.length == 1);
+      File indexDir = files[0];
+
+      // Convert the segment
+      File convertedSegmentDir = new File(tempDataDir, "convertedSegmentDir");
+      Preconditions.checkState(convertedSegmentDir.mkdir());
+      File convertedIndexDir = new File(convertedSegmentDir, segmentName);
+      convert(pinotTaskConfig, indexDir, convertedIndexDir);
+
+      // Tar the converted segment
+      File convertedTarredSegmentDir = new File(tempDataDir, "convertedTarredSegmentDir");
+      Preconditions.checkState(convertedTarredSegmentDir.mkdir());
+      File convertedTarredSegmentFile = new File(
+          TarGzCompressionUtils.createTarGzOfDirectory(convertedIndexDir.getPath(),
+              new File(convertedTarredSegmentDir, segmentName).getPath()));
+
+      // Check whether the task get cancelled before uploading the segment
+      if (_cancelled) {
+        throw new TaskCancelledException(
+            MinionConstants.ConvertToRawIndexTask.TASK_TYPE + " task on table: " + tableName + ", segment: "
+                + segmentName + " has been cancelled");
+      }
+
+      // Upload the converted tarred segment file with original segment crc and task type encoded in the http headers.
+      // Original segment crc is used to check whether the original segment get refreshed, so that the newer segment
+      // won't get override. We can decide how to handle the segment based on the task type on controller side.
+      // NOTE: even segment is not changed, still need to upload the segment to update the segment ZK metadata so that
+      // segment will not be submitted again
+      Header ifMatchHeader = new Header(HttpHeaders.IF_MATCH, originalSegmentCrc);
+      Header taskTypeHeader =
+          new Header(HttpHeaders.USER_AGENT, CommonConstants.Minion.HTTP_TASK_TYPE_HEADER_PREFIX + taskType);
```

Comment:
You may want to add in the default user agent parameter as well? I am fine adding things later as long as the parser at the other end accommodates for it.

Reply:
+1 on appending to the default user agent, so that we keep the version of the HTTP library used. The string should look like "ApacheHttp/1.2.3 Minion/1.0 SomeJob/2.1," so that we can detect it properly on the server side and apply workarounds as needed.

Reply:
Added